sigstore-java/src/main/java/dev/sigstore/bundle/BundleFactoryInternal.java

```diff
@@ -120,4 +144,79 @@ private static void addInclusionProof(
                     .collect(Collectors.toList()))
             .setCheckpoint(Checkpoint.newBuilder().setEnvelope(inclusionProof.getCheckpoint())));
   }
+
+  static KeylessSigningResult readBundle(Reader jsonReader) throws BundleParseException {
+    Bundle.Builder bundleBuilder = Bundle.newBuilder();
+    try {
+      JsonFormat.parser().merge(jsonReader, bundleBuilder);
+    } catch (IOException ioe) {
+      throw new BundleParseException("Could not read bundle json", ioe);
+    }
+    Bundle bundle = bundleBuilder.build();
+
+    var bundleEntry = bundle.getVerificationMaterial().getTlogEntries(0);
+    var bundleInclusionProof = bundleEntry.getInclusionProof();
+
+    ImmutableInclusionProof inclusionProof = null;
+    if (bundleEntry.hasInclusionProof()) {
+      inclusionProof =
+          ImmutableInclusionProof.builder()
+              .logIndex(bundleInclusionProof.getLogIndex())
+              .rootHash(Hex.toHexString(bundleInclusionProof.getRootHash().toByteArray()))
+              .treeSize(bundleInclusionProof.getTreeSize())
+              .checkpoint(bundleInclusionProof.getCheckpoint().getEnvelope())
+              .addAllHashes(
+                  bundleInclusionProof.getHashesList().stream()
+                      .map(ByteString::toByteArray)
+                      .map(Hex::toHexString)
+                      .collect(Collectors.toList()))
+              .build();
+    }
+
+    var verification =
+        ImmutableVerification.builder()
+            .signedEntryTimestamp(
+                Base64.getEncoder()
+                    .encodeToString(
+                        bundleEntry.getInclusionPromise().getSignedEntryTimestamp().toByteArray()))
+            .inclusionProof(Optional.ofNullable(inclusionProof))
+            .build();
+
+    var rekorEntry =
+        ImmutableRekorEntry.builder()
+            .integratedTime(bundleEntry.getIntegratedTime())
+            .logID(Hex.toHexString(bundleEntry.getLogId().getKeyId().toByteArray()))
+            .logIndex(bundleEntry.getLogIndex())
+            .body(
+                Base64.getEncoder()
+                    .encodeToString(bundleEntry.getCanonicalizedBody().toByteArray()))
+            .verification(verification)
+            .build();
+
+    if (bundle.getMessageSignature().getMessageDigest().getAlgorithm() != HashAlgorithm.SHA2_256) {
+      throw new BundleParseException("Cannot read message signatures that are not sha256");
```

Comment:
We should probably add actual algorithm name into the error message

Reply:
turns out there isn't much else (just sha2_256, unspecified and unrecognized), but added it in the message for the future.